File: lib/xero_cli/commands/contacts.ex
defmodule XeroCLI.Commands.Contacts do
  @moduledoc """
  Handles contact-related commands for Xero CLI.
  """

  @xero_api_base "https://api.xero.com/api.xro/2.0"

  def handle(:list, opts) do
    case get_valid_credentials() do
      {:ok, credentials} ->
        access_token = Map.get(credentials, "access_token")
        tenant_id = Map.get(credentials, "tenant_id")

        # Parse options
        page = get_option(opts, "--page", "1") |> String.to_integer()
        page_size = get_option(opts, "--page-size", "100") |> String.to_integer()

        # Build query parameters
        query_params =
          []
          |> maybe_add_param("page", page)
          |> maybe_add_param("pageSize", page_size)
          |> URI.encode_query()

        url = "#{@xero_api_base}/Contacts?#{query_params}"

        IO.puts("👥 Fetching contacts...\n")

        headers = [
          {"authorization", "Bearer #{access_token}"},
          {"xero-tenant-id", tenant_id},
          {"accept", "application/json"}
        ]

        case XeroCLI.HTTP.get(url, headers) do
          {:ok, %{status_code: 200, body: response_body}} ->
            case Jason.decode(response_body) do
              {:ok, data} ->
                contacts = get_in(data, ["Contacts"]) || []
                display_contacts(contacts)

              {:error, reason} ->
                IO.puts(:stderr, "❌ Failed to parse response: #{inspect(reason)}")
                System.halt(1)
            end

          {:ok, %{status_code: 401}} ->
            IO.puts(:stderr, "❌ Authentication failed. Please run 'xero auth login' again.")
            System.halt(1)

          {:ok, %{status_code: status, body: body}} ->
            IO.puts(:stderr, "❌ API request failed with status #{status}")
            IO.puts(:stderr, "Response: #{body}")
            System.halt(1)

          {:error, reason} ->
            IO.puts(:stderr, "❌ Request failed: #{inspect(reason)}")
            System.halt(1)
        end

      {:error, reason} ->
        IO.puts(:stderr, "❌ #{reason}")
        IO.puts("Please run 'xero auth login' first.")
        System.halt(1)
    end
  end

  def handle(:get, opts) do
    case opts do
      [contact_id | _rest] ->
        get_contact_by_id(contact_id)

      [] ->
        IO.puts(:stderr, "❌ Error: contact ID required")
        IO.puts("Usage: xero contacts get <contact_id>")
        System.halt(1)
    end
  end

  defp get_contact_by_id(contact_id) do
    case get_valid_credentials() do
      {:ok, credentials} ->
        access_token = Map.get(credentials, "access_token")
        tenant_id = Map.get(credentials, "tenant_id")

        url = "#{@xero_api_base}/Contacts/#{contact_id}"

        IO.puts("👥 Fetching contact #{contact_id}...\n")

        headers = [
          {"authorization", "Bearer #{access_token}"},
          {"xero-tenant-id", tenant_id},
          {"accept", "application/json"}
        ]

        case XeroCLI.HTTP.get(url, headers) do
          {:ok, %{status_code: 200, body: response_body}} ->
            case Jason.decode(response_body) do
              {:ok, data} ->
                contacts = get_in(data, ["Contacts"]) || []

                case contacts do
                  [contact | _] -> display_contact_detail(contact)
                  [] -> IO.puts("Contact not found.")
                end

              {:error, reason} ->
                IO.puts(:stderr, "❌ Failed to parse response: #{inspect(reason)}")
                System.halt(1)
            end

          {:ok, %{status_code: 404}} ->
            IO.puts(:stderr, "❌ Contact not found.")
            System.halt(1)

          {:ok, %{status_code: 401}} ->
            IO.puts(:stderr, "❌ Authentication failed. Please run 'xero auth login' again.")
            System.halt(1)

          {:ok, %{status_code: status, body: body}} ->
            IO.puts(:stderr, "❌ API request failed with status #{status}")
            IO.puts(:stderr, "Response: #{body}")
            System.halt(1)

          {:error, reason} ->
            IO.puts(:stderr, "❌ Request failed: #{inspect(reason)}")
            System.halt(1)
        end

      {:error, reason} ->
        IO.puts(:stderr, "❌ #{reason}")
        IO.puts("Please run 'xero auth login' first.")
        System.halt(1)
    end
  end

  defp get_valid_credentials do
    case XeroCLI.Config.get_credentials() do
      {:ok, credentials} ->
        # Check if token is expired
        if XeroCLI.OAuth.token_expired?(credentials) do
          IO.puts("🔄 Access token expired. Refreshing...")

          refresh_token = Map.get(credentials, "refresh_token")
          client_id = Map.get(credentials, "client_id")

          case XeroCLI.OAuth.refresh_token(refresh_token, client_id) do
            {:ok, new_token_data} ->
              # Update credentials with new token
              updated_credentials =
                credentials
                |> Map.put("access_token", Map.get(new_token_data, "access_token"))
                |> Map.put("refresh_token", Map.get(new_token_data, "refresh_token"))
                |> Map.put("expires_in", Map.get(new_token_data, "expires_in", 1800))
                |> Map.put("obtained_at", Map.get(new_token_data, "obtained_at"))

              case XeroCLI.Config.set_credentials(updated_credentials) do
                :ok ->
                  IO.puts("✓ Token refreshed\n")
                  {:ok, updated_credentials}

                {:error, reason} ->
                  {:error, "Failed to save refreshed credentials: #{reason}"}
              end

            {:error, reason} ->
              {:error, "Failed to refresh token: #{reason}"}
          end
        else
          {:ok, credentials}
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp get_option(opts, flag, default) do
    case Enum.find_index(opts, &(&1 == flag)) do
      nil -> default
      index -> Enum.at(opts, index + 1, default)
    end
  end

  defp maybe_add_param(params, _key, nil), do: params

  defp maybe_add_param(params, key, value) do
    [{key, value} | params]
  end

  defp display_contacts([]) do
    IO.puts("No contacts found.")
  end

  defp display_contacts(contacts) do
    IO.puts("Found #{length(contacts)} contact(s):\n")
    IO.puts(String.duplicate("=", 120))

    header =
      format_row([
        pad("Name", 30),
        pad("Email", 35),
        pad("Contact ID", 38),
        pad("Status", 15)
      ])

    IO.puts(header)
    IO.puts(String.duplicate("=", 120))

    Enum.each(contacts, fn contact ->
      name = Map.get(contact, "Name", "N/A")
      email = Map.get(contact, "EmailAddress", "N/A")
      contact_id = Map.get(contact, "ContactID", "N/A")
      status = Map.get(contact, "ContactStatus", "N/A")

      row =
        format_row([
          pad(name, 30),
          pad(email, 35),
          pad(contact_id, 38),
          pad(status, 15)
        ])

      IO.puts(row)
    end)

    IO.puts(String.duplicate("=", 120))
  end

  defp display_contact_detail(contact) do
    IO.puts("Contact Details:\n")
    IO.puts(String.duplicate("=", 80))

    name = Map.get(contact, "Name", "N/A")
    contact_id = Map.get(contact, "ContactID", "N/A")
    email = Map.get(contact, "EmailAddress", "N/A")
    status = Map.get(contact, "ContactStatus", "N/A")
    first_name = Map.get(contact, "FirstName", "N/A")
    last_name = Map.get(contact, "LastName", "N/A")

    IO.puts("Name:             #{name}")
    IO.puts("First Name:       #{first_name}")
    IO.puts("Last Name:        #{last_name}")
    IO.puts("Contact ID:       #{contact_id}")
    IO.puts("Email:            #{email}")
    IO.puts("Status:           #{status}")

    # Display addresses if available
    addresses = Map.get(contact, "Addresses", [])

    if length(addresses) > 0 do
      IO.puts("\nAddresses:")

      Enum.each(addresses, fn address ->
        type = Map.get(address, "AddressType", "N/A")
        line1 = Map.get(address, "AddressLine1", "")
        line2 = Map.get(address, "AddressLine2", "")
        city = Map.get(address, "City", "")
        region = Map.get(address, "Region", "")
        postal_code = Map.get(address, "PostalCode", "")
        country = Map.get(address, "Country", "")

        IO.puts("  #{type}:")
        if line1 != "", do: IO.puts("    #{line1}")
        if line2 != "", do: IO.puts("    #{line2}")

        location_parts =
          [city, region, postal_code, country]
          |> Enum.reject(&(&1 == ""))

        if length(location_parts) > 0 do
          IO.puts("    #{Enum.join(location_parts, ", ")}")
        end
      end)
    end

    # Display phones if available
    phones = Map.get(contact, "Phones", [])

    if length(phones) > 0 do
      IO.puts("\nPhones:")

      Enum.each(phones, fn phone ->
        type = Map.get(phone, "PhoneType", "N/A")
        number = Map.get(phone, "PhoneNumber", "N/A")
        IO.puts("  #{type}: #{number}")
      end)
    end

    IO.puts(String.duplicate("=", 80))
  end

  defp format_row(columns) do
    Enum.join(columns, " | ")
  end

  defp pad(str, width) do
    str = to_string(str)

    cond do
      String.length(str) > width ->
        String.slice(str, 0, width - 3) <> "..."

      String.length(str) < width ->
        String.pad_trailing(str, width)

      true ->
        str
    end
  end
end

File: cmd/xero/contacts.go
package main

import (
	"encoding/json"
	"fmt"
	"net/url"
	"strings"

	"github.com/msmithstubbs/xero-cli/internal/auth"
	"github.com/msmithstubbs/xero-cli/internal/ui"
	"github.com/msmithstubbs/xero-cli/internal/xero"
	"github.com/spf13/cobra"
)

const xeroAPIBase = "https://api.xero.com/api.xro/2.0"

var contactsCmd = &cobra.Command{
	Use:   "contacts",
	Short: "Manage contacts",
}

var contactsListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all contacts",
	RunE: func(cmd *cobra.Command, args []string) error {
		creds, err := auth.GetValidCredentials()
		if err != nil {
			return err
		}

		page, _ := cmd.Flags().GetInt("page")
		pageSize, _ := cmd.Flags().GetInt("page-size")

		params := url.Values{}
		if page > 0 {
			params.Set("page", fmt.Sprintf("%d", page))
		}
		if pageSize > 0 {
			params.Set("pageSize", fmt.Sprintf("%d", pageSize))
		}
		endpoint := fmt.Sprintf("%s/Contacts?%s", xeroAPIBase, params.Encode())

		if resolvedOutputFormat() == outputTable {
			fmt.Println("Fetching contacts...")
			fmt.Println()
		}

		headers, err := authHeaders(creds)
		if err != nil {
			return err
		}

		client := xero.NewClient(xeroAPIBase)
		status, body, err := client.Do("GET", endpoint, headers, nil)
		if err != nil {
			return internalError("request failed", err)
		}

		if status == 401 {
			return authenticationExpiredError()
		}
		if status < 200 || status >= 300 {
			return apiError(status, body)
		}

		var payload map[string]any
		if err := json.Unmarshal(body, &payload); err != nil {
			return parseResponseError(err)
		}

		contacts := getArray(payload, "Contacts")
		return emitData(payload, func() {
			displayContacts(contacts)
		})
	},
}

var contactsGetCmd = &cobra.Command{
	Use:   "get <contact_id>",
	Short: "Get a single contact by ID",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		contactID := args[0]
		creds, err := auth.GetValidCredentials()
		if err != nil {
			return err
		}

		endpoint := fmt.Sprintf("%s/Contacts/%s", xeroAPIBase, contactID)
		if resolvedOutputFormat() == outputTable {
			fmt.Printf("Fetching contact %s...\n\n", contactID)
		}

		headers, err := authHeaders(creds)
		if err != nil {
			return err
		}

		client := xero.NewClient(xeroAPIBase)
		status, body, err := client.Do("GET", endpoint, headers, nil)
		if err != nil {
			return internalError("request failed", err)
		}

		switch status {
		case 200:
			var payload map[string]any
			if err := json.Unmarshal(body, &payload); err != nil {
				return parseResponseError(err)
			}
			contacts := getArray(payload, "Contacts")
			if len(contacts) == 0 {
				return notFoundError("contact not found")
			}
			if contact, ok := contacts[0].(map[string]any); ok {
				return emitData(payload, func() {
					displayContactDetail(contact)
				})
			}
			return unexpectedResponseError()
		case 401:
			return authenticationExpiredError()
		case 404:
			return notFoundError("contact not found")
		default:
			return apiError(status, body)
		}
	},
}

func init() {
	contactsCmd.AddCommand(contactsListCmd)
	contactsCmd.AddCommand(contactsGetCmd)
	contactsListCmd.Flags().Int("page", 1, "Page number for pagination")
	contactsListCmd.Flags().Int("page-size", 100, "Number of items per page")
}

func displayContacts(items []any) {
	if len(items) == 0 {
		fmt.Println("No contacts found.")
		return
	}

	fmt.Printf("Found %d contact(s):\n", len(items))
	fmt.Println()
	ui.PrintHeaderLine(120)
	header := ui.FormatRow(
		ui.Pad("Name", 30),
		ui.Pad("Email", 35),
		ui.Pad("Contact ID", 38),
		ui.Pad("Status", 15),
	)
	fmt.Println(header)
	ui.PrintHeaderLine(120)

	for _, item := range items {
		contact, ok := item.(map[string]any)
		if !ok {
			continue
		}
		name := stringValue(contact, "Name", "N/A")
		email := stringValue(contact, "EmailAddress", "N/A")
		contactID := stringValue(contact, "ContactID", "N/A")
		status := stringValue(contact, "ContactStatus", "N/A")

		row := ui.FormatRow(
			ui.Pad(name, 30),
			ui.Pad(email, 35),
			ui.Pad(contactID, 38),
			ui.Pad(status, 15),
		)
		fmt.Println(row)
	}

	ui.PrintHeaderLine(120)
}

func displayContactDetail(contact map[string]any) {
	fmt.Println("Contact Details:")
	fmt.Println()
	ui.PrintHeaderLine(80)

	fmt.Printf("Name:             %s\n", stringValue(contact, "Name", "N/A"))
	fmt.Printf("First Name:       %s\n", stringValue(contact, "FirstName", "N/A"))
	fmt.Printf("Last Name:        %s\n", stringValue(contact, "LastName", "N/A"))
	fmt.Printf("Contact ID:       %s\n", stringValue(contact, "ContactID", "N/A"))
	fmt.Printf("Email:            %s\n", stringValue(contact, "EmailAddress", "N/A"))
	fmt.Printf("Status:           %s\n", stringValue(contact, "ContactStatus", "N/A"))

	addresses := getArray(contact, "Addresses")
	if len(addresses) > 0 {
		fmt.Println("\nAddresses:")
		for _, addressItem := range addresses {
			address, ok := addressItem.(map[string]any)
			if !ok {
				continue
			}
			addrType := stringValue(address, "AddressType", "N/A")
			line1 := stringValue(address, "AddressLine1", "")
			line2 := stringValue(address, "AddressLine2", "")
			city := stringValue(address, "City", "")
			region := stringValue(address, "Region", "")
			postal := stringValue(address, "PostalCode", "")
			country := stringValue(address, "Country", "")

			fmt.Printf("  %s:\n", addrType)
			if line1 != "" {
				fmt.Printf("    %s\n", line1)
			}
			if line2 != "" {
				fmt.Printf("    %s\n", line2)
			}
			location := strings.TrimSpace(strings.Join(filterEmpty([]string{city, region, postal, country}), ", "))
			if location != "" {
				fmt.Printf("    %s\n", location)
			}
		}
	}

	phones := getArray(contact, "Phones")
	if len(phones) > 0 {
		fmt.Println("\nPhones:")
		for _, phoneItem := range phones {
			phone, ok := phoneItem.(map[string]any)
			if !ok {
				continue
			}
			typeValue := stringValue(phone, "PhoneType", "N/A")
			number := stringValue(phone, "PhoneNumber", "N/A")
			fmt.Printf("  %s: %s\n", typeValue, number)
		}
	}

	ui.PrintHeaderLine(80)
}
